src/Stratis.Bitcoin/Base/BackgroundWork/BackgroundWorkProvider.cs

```diff
@@ -0,0 +1,214 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+using Stratis.Bitcoin.Signals;
+using Stratis.Bitcoin.Utilities;
+
+namespace Stratis.Bitcoin.Base.BackgroundWork
+{
+    /// <summary>
+    /// Provides functionality for creating and tracking operations that happens in background.
+    /// </summary>
+    public partial class BackgroundWorkProvider : IBackgroundWorkProvider
+    {
+        private const int DefaultLoopRepeatInterval = 1000;
+
+        private object lockAsyncDelegates;
+        private object lockAsyncLoops;
+
+        /// <summary>
+        /// Holds a list of currently running async delegates or delegates that stopped because of unhandled exceptions.
+        /// Protected by <see cref="lockAsyncDelegates"/> lock
+        /// </summary>
+        private Dictionary<IAsyncDelegateWorker, AsyncTaskInfo> asyncDelegateWorkers;
+
+        /// <summary>
+        /// Holds a list of currently running async loops or loops that stopped because of unhandled exceptions.
+        /// Protected by <see cref="lockAsyncLoops"/> lock
+        /// </summary>
+        private Dictionary<IAsyncLoop, AsyncTaskInfo> asyncLoops;
+
+        private ILoggerFactory loggerFactory;
+        private ILogger logger;
+        private ISignals signals;
+        private readonly INodeLifetime nodeLifetime;
+        private readonly IAsyncLoopFactory asyncLoopFactory;
```

Comment:
`IAsyncLoopFactory` can go away? we move its logic in to `AsyncProvider`?

Reply:
that's the plan, but actually there is a scenario where we are using RunUntil that I'm not sure if implement in our provider or not so atm I'm leaving it here until we start using this provider (another PR?)

Reply:
after reviewing it, yes, actually I can remove it (and I did it)